add usaco-1206 solution (#2676)
cpinitiative/usaco-guide: solutions/usaco-1206.mdx

---
id: usaco-1206
source: USACO Silver 2022 February
title: Redistributing Gifts
author: Chongtian Ma
---

[Official Analysis](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)

## Explanation

Similar to the official analysis, we created a directed graph with edges from $i$ to $j$ if cow $i$ prefers gift $j$ more compared to his own. 
If cow $i$ is able to receive a gift where it has an outgoing edge, then the solution will always be more optimal. 
Remember to also add an edge between cow $i$ and $i$ in the case where it cannot get a more preferred gift.

Then, I separated the graph into strongly connected components. If $i$ is in a strongly connected component, this means that cow $i$ can receive
all gifts in the strongly connected component through a series of reassignments and is guaranteed to have one or more preferred gift (including itself) based 
on SCC definition. Obviously you only want to assign cow $i$ with a more preferred gift.
## Implementation

> I used Kosaraju's algorithm to find all strongly connected components in the graph. At the end, I just went down cow $i$'s preference list and assigned the first gift belonging in the SCC.

Reviewer: don't use first person

Reviewer: not addressed

Author: fixed @SansPapyrus683 
By the way is it also okay if I use "we"?

Time complexity is $O(M)$ where $M$ is the number of edges as mentioned at the bottom of the official analysis.

<LanguageSection>
<CPPSection>
```cpp
#include <bits/stdc++.h>
using namespace std;

> #define all(x) x.begin(), x.end()

Reviewer: please don't use macros

> const int N = 505;

Reviewer: set N = 500 instead of 505

Author: Well i used one indexed so ill set it to 501 if thats ok.

Reviewer: sure

// for kosaraju purposes
vector<int> graph[N], revgraph[N]; // revgraph is the reversed graph
vector<bool> vis(N), vis2(N);
vector<int> path; 

// stores all nodes in a SCC
set<int> comp;

void dfs(int node){
    vis[node] = true;
    for(int i: graph[node]){
        if(!vis[i]){
            dfs(i);
        }
    }
    path.push_back(node);
}

void dfs2(int node){
    vis2[node] = true;
    for(int i: revgraph[node]){
        if(!vis2[i]){
            dfs2(i);
        }
    }
    comp.insert(node);
}
signed main() {
    ios::sync_with_stdio(0);
    cin.tie(0);
    int n; cin>>n;
    for(int i = 1; i <= n; i++){
        // NOTE that graph array serves as its preference list and an adjacency list of the directed graph.

>         graph[i].resize(n);

Reviewer: use tabs (please, _please_ read the style guide when implementing a PR)

Author: I'm sorry but I believe I am using tabs? I pasted this code into CLion (which should auto format the code) before I pasted it here. Is it faulty or can you point out where I am not using tabs? Thanks @SansPapyrus683

Author: fixed

        for(int j = 0; j < n; j++){
            cin>>graph[i][j];
        }
        while(graph[i].back() != i){

>             graph[i].pop_back(); // we don't want edges to non-preferred gifts so remove all gifts after its original assignment

Reviewer: over 80 chars

---
id: usaco-1206
source: USACO Silver 2022 February
title: Redistributing Gifts
author: Chongtian Ma
---

[Official Analysis](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)

> ## Explanation

Reviewer 2: Don't talk in first person. Talk in third person like all editorials do (change I to we).

Author: fixed in the explanation part, but i kept the I in my implementation expalanation

> Similar to the official analysis, I created a directed graph with edges from $i$ to $j$ if cow $i$ prefers gift $j$ more compared to his own. 

Reviewer 2: ```suggestion
Similar to the official analysis, I constructed directed edges from $i$ to $j$ if cow $i$ prefers gift $j$ to gift $i$.
```

> This is because it is always more optimal for cow $i$ to receive one of these gifts. In other words, if cow $i$ is able to receive a gift where

Reviewer 2: Isn't this kind of obvious?

Author: removed that sentence, yeah i agree.

it has an outgoing edge, then the solution will always be more optimal. Remember to also add an edge between cow $i$ and $i$ in the case where
it cannot get a more preferred gift.

> Then, I broke up the graph into strongly connected components. If $i$ is in a strongly connected component, this means that cow $i$ can receive

Reviewer 2: ```suggestion
Then, I separated the graph into strongly connected components. If $i$ is in a strongly connected component, this means that cow $i$ can receive
```

> all gifts in the strongly connected component through a series of reassignments and is guaranteed to have one or more preferred gift (including itself).

Reviewer 2: ```suggestion
all gifts in the strongly connected component through a series of reassignments and is guaranteed to have one or more preferred gifts (including itself).
```
maybe mention something about a cycle here?

Author: I believe saying its an SCC is kind of self implied, which i personally think its a bit redundant. If you insist I can add it.

> This is because if cow $i$ and cow $j$ is in the same SCC, then it is guaranteed cow $i$ can receive gift $j$ through a series of traversals (proof by

Reviewer 2: ```suggestion
This is because if cow $i$ and cow $j$ are in the same strongly connected component, then it is guaranteed cow $i$ can receive gift $j$ through a series of traversals (proof by
```
You basically said the same thing as the previous sentence. Just mention cycles in that previous sentence and that'd make the explanation more concise...

Author: i removed the second sentence. I think mentioning cycles is a bit redundant though.

SCC definition). Obviously you only want to assign cow $i$ with a more preferred gift.
## Implementation

> I used Kosaraju's algorithm to find all SCCs in the graph. At the end, I just assigned cow $i$ with the first gift in the SCC that it has preference to.

Reviewer 2: ```suggestion
I used Kosaraju's algorithm to find all strongly connected components in the graph.
```
What does "At the end, I just assigned cow $i$ with the first gift in the SCC that it has preference to." mean?

Author: fixed, see if its better now.

> Time complexity is $O(M)$ where $M$ is the number of edges as mentioned at the bottom of the official analysis.

Reviewer 2: ```suggestion
**Time Complexity:** $\mathcal{O}(N^3)$
```

Author: I dont know if this needs to be changed, can another admin confirm.

Reviewer 2: @SansPapyrus683

Reviewer: confirmed

<LanguageSection>
<CPPSection>
```cpp

> #include <bits/stdc++.h>

Reviewer 2: Use tabs instead of spaces to indent

using namespace std;
#define all(x) x.begin(), x.end()
const int N = 505;

// for kosaraju purposes

> vector<vector<int>> graph(N), revgraph(N);

Reviewer 2: vector<int> graph[N], revgraph[N];

Also you should probably make a comment describing that revgraph is the reversed graph

Author: Well, its a essential part of kosaraju and I think the name revgraph is self implied. I'll still add it though.

        graph[i].resize(n);
        for(int j = 0; j < n; j++){
            cin>>graph[i][j];
        }
        while(graph[i].back() != i){
            graph[i].pop_back(); // we don't want edges to non-preferred gifts so remove all gifts after its original assignment
        }
        for(int j: graph[i]){
            revgraph[j].push_back(i);
        }
    }
    
    // classic kosaraju implementation
    for(int i = 1; i <= n; i++){
        if(!vis[i]){
            dfs(i);
        }
    }                    
    reverse(all(path));     
                         
    vector<int> ans(n+1);
    for(int i: path){
        if(!vis2[i]){
            dfs2(i);
            // loop over all cows in the SCC and its preference list. If we find a gift in its preference and also in the SCC, just assign it.
            for(int j: comp){
                for(int k: graph[j]){
                    if(comp.count(k)){
                        ans[j] = k;
                        break;
                    }
                }
            }

            comp.clear();
        }
    }
    for(int i = 1; i <= n; i++){
        cout<<ans[i]<<endl;
    }

> 

Reviewer 2: why so much extra space?

}

signed main() {
    ios::sync_with_stdio(0);
    cin.tie(0);

>     solve();

Reviewer 2: Put everything from ```solve()``` in the main method...

---
id: usaco-1206
source: USACO Silver 2022 February
title: Redistributing Gifts
author: Chongtian Ma
---

[Official Analysis](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)

## Explanation

Similar to the official analysis, we can construct a directed edge from $i$ to $j$ if cow $i$ prefers gift $j$ to gift $i$.

> If cow $i$ is able to receive a gift where it has an outgoing edge, then the solution will always be more optimal. 

Reviewer: > more optimal

more optimal than what?

> If cow $i$ is able to receive a gift where it has an outgoing edge, then cow $i$ has received the most optimal gift it can be reassigned.

Reviewer: what does this mean?

Reviewer: also nothing's mentioned about cycles- is that intentional?

Remember to also add an edge from cow $i$ to itself in case it cannot get a better gift.

> Then, we can separate the graph into strongly connected components (SCCs). If $i$ is in an SCC, this means that cow $i$ can receive

Reviewer: what's the intuition behind this?

Author: It’s just implementation beauty since running kosaraju you are guaranteed to least one node with an outgoing edge (itself). Should i mention it? I didn’t becuase i thought it was kind of redundant.

Author: Also it helps in the case where it has the original gift already at the most preferred gift.

all gifts in the SCC through a series of reassignments by definition of an SCC. It is also guaranteed to receive one 
or more preferred gifts (including the original assignment) since we are only adding these gifts as outgoing edges. If a cow does not belong in 

> an SCC, it cannot receive a more optimal reassignment.

Reviewer: > It is also guaranteed to receive one or more preferred gifts (including the original assignment) since we are only adding these gifts as outgoing edges.

what does this mean

Author: It is kind of redundant since if its in an SCC it has to have at least one outgoing edge to. Is it fine if I just remove that sentence?

Reviewer: up to you

Author: is it better now?

## Implementation

> Kosaraju's algorithm can be used to find all SCCs in the graph. At the end, it is optimal to just go down cow $i$'s preference list and assign the first gift belonging in the SCC.

Reviewer: add a link to this website here: https://en.wikipedia.org/wiki/Kosaraju's_algorithm

**Time Complexity:** $\mathcal{O}(N^3)$

<LanguageSection>
<CPPSection>
```cpp
#include <bits/stdc++.h>

using namespace std;
const int N = 501;

// The adjacency lists for the graph (revgraph is reversed)
vector<int> graph[N], revgraph[N];
vector<bool> vis(N), vis2(N);
vector<int> path;

// stores all nodes in a SCC
set<int> comp;

void dfs(int node) {
	vis[node] = true;
	for (int i: graph[node]) {
		if (!vis[i]) {
			dfs(i);
		}
	}
	path.push_back(node);
}

void dfs2(int node) {
	vis2[node] = true;
	for (int i: revgraph[node]) {
		if (!vis2[i]) {
			dfs2(i);
		}
	}
	comp.insert(node);

> }

Reviewer: wth do these two methods do
`dfs` and `dfs2` give no insight into their functionality

Author: Well its an part of kosaraju. Should i explain why it works in kosaraju or relate it to the wikipedia article?

Reviewer: both of them do kosaraju's? in that case yeah mention it in one comment
`// DFS functions for Kosaraju's`

---
id: usaco-1206
source: USACO Silver 2022 February
title: Redistributing Gifts
author: Chongtian Ma
---

[Official Analysis (C++)](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)

## Solution 1

> See [Official Analysis](http://www.usaco.org/current/data/sol_prob1_silver_feb22.html)

Reviewer: ```suggestion
See the official analysis.
```
